src/main/java/spoon/reflect/visitor/chain/CtQuery.java

```diff
@@ -23,59 +23,91 @@
+	 * The matched child element for which (filter.matches(element)==true) are sent to the next query step.
+	 * Essentially the same as {@link CtElement#getElements(Filter)} but more powerful, because it
+	 * can be chained with other subsequent queries.
+	 *
+	 * Note: the input element (the root of the query, `this` if you're in {@link CtElement}) is also checked and may thus be also sent to the next step.
+	 * The elements which throw {@link ClassCastException} during {@link Filter#matches(CtElement)}
+	 * are considered as **not matching**, ie. are excluded.
+	 *
+	 * @param filter used to filter scanned children elements of the AST tree
+	 * @return a new Query
+	 */
+	@Override
+	<R extends CtElement> CtQuery filterChildren(Filter<R> filter);
+
+
+	/**
+	 * Query elements based on a function, the behavior depends on the return type of the function.
+	 * <table>
+	 * <tr><td><b>Return type of `function`</b><td><b>Behavior</b>
+	 * <tr><td>{@link Boolean}<td>Select elements if thereturned value of `function` is true (as for {@link Filter}).
+	 * <tr><td>? extends {@link Object}<td>Send the returned value of `function` to the next step
+	 * <tr><td>{@link Iterable}<td>Send each item of the collection to the next step
+	 * <tr><td>{@link Object[]}<td>Send each item of the array to the next step
+	 * </table><br>
+	 *
+	 * @param function a Function with one parameter of type I returning a value of type R
+	 * @return a new query object
+	 */
+	@Override
+	<I, R> CtQuery map(CtFunction<I, R> function);
+
+	/**
+	 * Sets (binds) the input of the query, the search will be done on children of that element.
+	 * If the query is created by {@link CtElement#map} or {@link CtElement#filterChildren(Filter)},
 	 * then the query is already bound to this element.
-	 * A new call of {@link #setInput(Object...)} will reset the current binding ans use the new one.
+	 * A new call of {@link #setInput(Object...)} will is always possible, it resets the current binding ans sets the new one.
 	 *
 	 * @param input
 	 * @return this to support fluent API
 	 */
 	<T extends CtQuery> T setInput(Object... input);
 
 	/**
-	 * actually evaluates the query and for each produced outputElement calls `consumer.accept(outputElement)`
+	 * Actually evaluates the query and for each produced output element of the last step,
+	 * calls `consumer.accept(outputElement)`.
+	 *
+	 * This avoids to create useless intermediate lists.
+	 *
      * @param consumer The consumer which accepts the results of the query
 	 */
 	<R> void forEach(CtConsumer<R> consumer);
+
 	/**
-	 * actually evaluates the query and returns all the produced elements collected in a List
+	 * Actually evaluates the query and returns all the elements produced in the last step.
 	 * @return the list of elements collected by the query.
 	 * @see #forEach(CtConsumer) for an efficient way of manipulating the elements without creating an intermediate list.
 	 */
 	<R extends Object> List<R> list();
+

```

Comment:
What about some note like:
You can use any item type of the List. The compiler nor the query engine will check it. So the wrong chosen List type can cause ClassCastException